src/hubAuth.ts

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (c) 2020, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import * as path from 'path';
+import * as os from 'os';
+import { fs } from '@salesforce/core';
+import * as shell from 'shelljs';
+import { debug } from 'debug';
+
+// this seems to be a known eslint error for enums
+// eslint-disable-next-line no-shadow
+enum AuthStrategy {
+  JWT = 'JWT',
+  AUTH_URL = 'AUTH_URL',
+  REUSE = 'REUSE',
+  NONE = 'NONE',
+}
+
+export const authFromStubbedHome = (homeDir: string): void => {
+  const logger = debug('testkit:authFromStubbedHome');
+  if (getAuthStrategy() === AuthStrategy.JWT && process.env.TESTKIT_JWT_KEY) {
+    logger('trying jwt auth');
+    const jwtKey = path.join(homeDir, 'jwtKey');
+    fs.writeFileSync(jwtKey, process.env.TESTKIT_JWT_KEY);
+
+    shell.exec(
+      `sfdx auth:jwt:grant -d -u ${process.env.TESTKIT_HUB_USERNAME} -i ${
+        process.env.TESTKIT_JWT_CLIENT_ID
+      } -f ${jwtKey} -r ${process.env.TESTKIT_HUB_INSTANCE || 'https://login.salesforce.com'}`,
+      { silent: true }
+    );
+    return;
+  }
+  if (getAuthStrategy() === AuthStrategy.AUTH_URL && process.env.TESTKIT_AUTH_URL) {
+    logger(`trying url based auth using ${process.env.TESTKIT_AUTH_URL} for dir ${homeDir}`);
+
+    const tmpUrl = path.join(homeDir, 'tmpUrl');
+    fs.writeFileSync(tmpUrl, process.env.TESTKIT_AUTH_URL);
```

Comment:
Same question. Why the need to save what could be a secure value?

Reply:
The auth command only accepts a file parameter, right?  It has to become a file at some point

Reply:
@mshanemc thank you, was not aware, but same applies here. How do we need to respond to vault capabilities across CI/CD systems?

Reply:
I've primarily used github actions and circle.  They read from env vars, and that seems like a pretty universal pattern, plus this satisifies local dev.

Can you say more about what you're thinking about?

Reply:
Here are the [Jenkins docs](https://www.jenkins.io/doc/book/using/using-credentials/) on its credential vault for to review.

Primarily it deals with strings of various purposes, but the one unique is the ability to upload a file into the store. When that credential is called out, Jenkins makes it available as a file.

If CircleCi secret vault is only string then we will need the facilities to store values into files for our commands to consume. But we should be aware of the differences between CI/CD systems in order to meet the needs of customers that may not be using circleci.

We could move forward with a string only interface for well-known secrets, but that does impose our will on our customers.

Reply:
I found this [article](https://support.circleci.com/hc/en-us/articles/360006717953-Storing-Secret-Files-certs-etc-) for Circle that talks about storing files.

Reply:
I think I would rather see the CI/CD environment supply the secret in the form required, rather than make it a responsibility of our code. So, for JWT key file, regardless of vault, the test env presents a file to the test. In Jenkins that would be through secret file credential and in CircleCi, the job would take the env var and echo the contents into a file before the tests are kicked off.

Reply:
besides circle...github actions and herokuCI are also only-strings.  It's a pretty common pattern.  It's cool that Jenkins has a file vault.

Github/Circle have an option to encrypt and store the key in the repo (I don't like that pattern where users would need the key).

I want testkit to reduce setup complexity, which is why the code is simplifying this, to say "just use these envVar" vs. "choose a CI and figure out how it wants you to get envs into files that our auth commands require and then also make sure your CI puts them in a location specified by another envVar that the commands reference"

Reply:
I'll create a WI for file locations as an envVar

Reply:
Its really not that complicated,

```
withCredentials([file(credentialsId: 'JWT_VAULT_FILE_ID', variable: 'JWT_FILE')]) {
    // now env var JWT_FILE is populated with the file's location
}
```

Reply:
Our circleci examples actually do one of the methods mentioned in the link Pete attached. It is a lot more complicated and a lot more setup. We already got approved to store secret files in circleci envvars, so I think we need to add a story to update our circleci examples to do the same thing we do within our own circleci jobs. 

Also, I think a separate story for file locations as an envvar is a good compramise. `TESTKIT_AUTH_URL` v.s `TESTKIT_AUTH_URL_FILE_PATH` or `TESTKIT_JWT_KEY` v.s `TESTKIT_JWT_KEY_FILE_PATH`

Reply:
W-8898036

Reply:
@shetzel `TESTKIT_AUTH_URL` is an env var that gets written to a file at test-time.

Not everyone in every system has access to the entire FS (think heroku dynos, or this scenario https://github.com/forcedotcom/cli/issues/848) to write files to a place where multiple tests can read them.

And if we require it to be a file, someone, somewhere has to somehow get that file into a known place that tests can rely on.  So now that's 2 required env vars (one for the actual secret to write to, and another for where to write it to/read if from)